### PR TITLE
FRONTEND-5940 :: Feature Internal :: Provide the status of the actions taken on harmony-typescript from the RW session

### PR DESCRIPTION
**Pull request by orioljp**
**Merge / Pull request information:**
Test for Cache Repository

* Asana task link: 
https://app.asana.com/0/1109863238977521/1202727845355940

### CODE REVIEW

**Comment by orioljp** [comment]
@migueltorcha , this PR contains two tests. Could you give us your opinion about the one for the CacheRepository? We already have another PR for the InMemoryDataSource in another PR. I'll delete it from this one.

**Comment by migueltorcha** [comment]
> @migueltorcha , this PR contains two tests. Could you give us your opinion about the one for the CacheRepository? We already have another PR for the InMemoryDataSource in another PR. I'll delete it from this one.

I would say that test covers all the expected behaviours in the code, so it looks nice for me. Anyway I have some proposals:

- I would try to use a more standard naming for the tests. Take a look to this link:
https://medium.com/@stefanovskyi/unit-test-naming-conventions-dd9208eadbea
My personal pick of all those options is this -> "should HAVE_EXPECTED_BEHAVIOUR when RACE_CONDITION"
This option have a natural language and it is not couple to any function name, so if you refactor your code, it will not become obsolete.

- Probably you cannot do it because of the language, no idea, but I'll try to follow the given-when-then structure. For example:
```

        it('should return the expected value when it is saved', () => {
            // given
            expect.assertions(1);
            const query = getDefaultKeyQuery();
            const operation = getDefaultOperation();
            const book = getDefaultBook();
            repository.put(book, query, operation);
            
            // when
            const result = repository.get(query, operation)

            // then
            return expect(result).resolves.toEqual(book);
        });
```

No need to keep the comments, this is just to make you aware of the 3 sections in the structure.

**Review by doup — CHANGES_REQUESTED** [review]

**Comment by orioljp** [comment]
I refactored the code that @Alex-DA provided to test the CacheRepository and following Miguel instructions.

@doup could you check it?

**Review by joselufo — APPROVED** [review]

File: packages/core/tests/repository/cache.repository.test.ts [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+            const book = getDefaultBook();
+            const keyQuery = getDefaultKeyQuery();
+            const notSupportedOperation = getNotSupportedOperation();
+
+            const result = repository.put(book, keyQuery, notSupportedOperation);
+
+            expect(result).rejects.toThrowError(OperationNotSupportedError);
+        });
+
+        it('should return the given value when this is inserted', async () => {
```
Comment (doup) [inline packages/core/tests/repository/cache.repository.test.ts:102]:
```suggestion
        it('should return the given value on insertion', async () => {
```

File: packages/core/tests/repository/cache.repository.test.ts [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+            const book = getDefaultBook();
+            const keyQuery = getDefaultKeyQuery();
+            const operation = getDefaultOperation();
+
+            const result = await repository.put(book, keyQuery, operation);
+
+            expect(result).toBe(book);
+        });
+
+        it('should override the value when an already existing key is given', async () => {
```
Comment (doup) [inline packages/core/tests/repository/cache.repository.test.ts:112]:
```suggestion
        it('should override the value if the key already exists', async () => {
```

File: packages/core/tests/repository/cache.repository.test.ts [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        it('should throw an error when a non-supported operation is given', () => {
+            const keyQuery = getDefaultKeyQuery();
+            const notSupportedOperation = getNotSupportedOperation();
+
+            const result = repository.delete(keyQuery, notSupportedOperation);
+
+            expect(result).rejects.toThrowError(OperationNotSupportedError);
+        });
+
+        it('should return a value for a given key when the delete was performed on another key', async () => {
```
Comment (doup) [inline packages/core/tests/repository/cache.repository.test.ts:148]:
```suggestion
        it(`shouldn't affect non-related keys (deletion doesn't have side effects)`, async () => {
```

File: packages/core/tests/repository/cache.repository.test.ts [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+        it('should throw an error when a non-supported operation is given', () => {
+            const keyQuery = getDefaultKeyQuery();
+            const notSupportedOperation = getNotSupportedOperation();
+
+            const result = repository.get(keyQuery, notSupportedOperation);
+
+            expect(result).rejects.toThrow(OperationNotSupportedError);
+        });
+
+        it('should return undefined when the key is not found', async () => {
```
Comment (doup) [inline packages/core/tests/repository/cache.repository.test.ts:48]:
```suggestion
        it('should return `undefined` if the key is not found', async () => {
```

File: packages/core/tests/repository/cache.repository.test.ts [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+            const bookTwo = getRandomBook();
+
+            await repository.put(bookOne, query, operation);
+            await repository.put(bookTwo, query, operation);
+            const result = await repository.get(query, operation);
+
+            expect(result).toBe(bookTwo);
+        });
+
+        it('should not update the cache when the MainSyncOperation fails', async () => {
```
Comment (doup) [inline packages/core/tests/repository/cache.repository.test.ts:125]:
```suggestion
        it("shouldn't update the cache when the `MainSyncOperation` fails", async () => {
```

File: packages/core/tests/repository/cache.repository.test.ts [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+            const syncResult = repository.put(book, query, operation)
+            const cacheResult = repository.get(getDefaultKeyQuery(), getCacheOperation());
+
+            await expect(syncResult).rejects.toThrowError(QueryNotSupportedError);
+            await expect(cacheResult).rejects.toThrowError(NotFoundError);
+        });
+    });
+
+    describe('delete', () => {
+        it('should throw an error when a non-supported operation is given', () => {
```
Comment (doup) [inline packages/core/tests/repository/cache.repository.test.ts:139]:
```suggestion
        it('should throw an error when an unsupported operation is given', () => {
```

File: packages/core/tests/repository/cache.repository.test.ts [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+            mainDataSource,
+            cacheDataSource,
+            cacheDataSource,
+            cacheDataSource,
+            validator,
+        );
+    });
+
+    describe('get', () => {
+        it('should throw an error when a non-supported operation is given', () => {
```
Comment (doup) [inline packages/core/tests/repository/cache.repository.test.ts:39]:
```suggestion
        it('should throw an error when an unsupported operation is given', () => {
```

File: packages/core/tests/repository/cache.repository.test.ts [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+
+            await repository.put(book, query, operation);
+            const result = await repository.get(query, operation);
+
+            expect(result).toBe(book);
+        });
+    });
+
+    describe('put', () => {
+        it('should throw an error when a non-supported operation is given', async () => {
```
Comment (doup) [inline packages/core/tests/repository/cache.repository.test.ts:92]:
```suggestion
        it('should throw an error when an unsupported operation is given', async () => {
```

File: packages/core/tests/repository/cache.repository.test.ts [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+            const book = getDefaultBook();
+
+            await repository.put(book, queryExisting, operation);
+            await repository.delete(queryNotExisting, operation);
+            const result = await repository.get(queryExisting, operation);
+
+            expect(result).toBe(book);
+        });
+
+        it('should return undefined when a value is deleted', async () => {
```
Comment (doup) [inline packages/core/tests/repository/cache.repository.test.ts:161]:
```suggestion
        it('should return `undefined` when a value is deleted', async () => {
```

**Review by doup — COMMENTED** [review]
I've added few suggestion for the name of the tests.

**Review by doup — APPROVED** [review]
Otherwise it looks good. I'm approving anyway, I prefer the suggested test names, but up to you.

**Review by lucianosantana — APPROVED** [review]

File: packages/core/tests/repository/cache.repository.test.ts [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+        it('should return undefined when the key is not found', async () => {
+            const nonExistingKeyQuery = getMissingKeyQuery()
+            const operation = getDefaultOperation();
+
+            const result = await repository.get(nonExistingKeyQuery, operation);
+
+            expect(result).toBe(undefined);
+        });
+
+        it('should return the expected value when an existing is given', async () => {
```
[the remark this answers is not in the document]
Comment (orioljp) [reply packages/core/tests/repository/cache.repository.test.ts:57]: Following what Miguel said:
_My personal pick of all those options is this -> "should HAVE_EXPECTED_BEHAVIOUR when RACE_CONDITION"_

File: packages/core/tests/repository/cache.repository.test.ts (hunk 1 above)
Comment (orioljp) [reply packages/core/tests/repository/cache.repository.test.ts:102]: Following what Miguel said:
My personal pick of all those options is this -> "should HAVE_EXPECTED_BEHAVIOUR when RACE_CONDITION"

File: packages/core/tests/repository/cache.repository.test.ts (hunk 2 above)
Comment (orioljp) [reply packages/core/tests/repository/cache.repository.test.ts:112]: Following what Miguel said:
My personal pick of all those options is this -> "should HAVE_EXPECTED_BEHAVIOUR when RACE_CONDITION"

File: packages/core/tests/repository/cache.repository.test.ts (hunk 3 above)
Comment (orioljp) [reply packages/core/tests/repository/cache.repository.test.ts:148]: Following what Miguel said:
My personal pick of all those options is this -> "should HAVE_EXPECTED_BEHAVIOUR when RACE_CONDITION"

**Comment by orioljp** [comment]
@doup I added most of your suggestions, thanks for reviewing the texts. I'm not changing things like "should not" into shouldn't because I preffer to right in a formal way.

File: packages/core/tests/repository/cache.repository.test.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (doup) [reply packages/core/tests/repository/cache.repository.test.ts:57]: For me the wording sounds just wrong; what business sense does have `expected value`? That doesn't explain anything. Similarly `when an existing is given`, existing what?

If you want to maintain the `should X when Y`, then maybe:
`should return the cached value when an existing key is given`

File: packages/core/tests/repository/cache.repository.test.ts (hunk 1 above)
Comment (doup) [reply packages/core/tests/repository/cache.repository.test.ts]: In this case `on` also denotes an event; like when you say `onClick` you could call it `whenTheUserClicks`; I think is synonym.

File: packages/core/tests/repository/cache.repository.test.ts (hunk 2 above)
Comment (doup) [reply packages/core/tests/repository/cache.repository.test.ts:112]: `should override the value when the key already exists`

File: packages/core/tests/repository/cache.repository.test.ts (hunk 3 above)
Comment (doup) [reply packages/core/tests/repository/cache.repository.test.ts:148]: Is super weird. The `when` denotes causality, no? So… delete was performed on another key, which triggered a value return for the given key?

What about? `should not affect other cache entries when an entry is deleted`. This reads better: an entry was deleted which didn't affect other cache entries.

**Pull request merged**